### PR TITLE
Add inner topic protection

### PR DESCRIPTION
**Pull request by hangc0276**
### Motivation
For inner topic, such as, __consumer_offsets, __transaction_stats, and system topic, such as, __change_event, they are expose to all users and can produce / consumer message to / from those topic, which will cause the cluster crash.

### Changes
1. add produce protection for inner topic on KOP side.

TODO
1. for consume protection, it's in testing.

### CODE REVIEW

**Comment by BewareMyPower** [comment]
Just a doubt about your TODO: do we need to protect consuming? If I didn't remember wrong, Kafka allows consumers to read `__consumer_offsets` and some admin tools are implemented by this way.

**Comment by hangc0276** [comment]
> Just a doubt about your TODO: do we need to protect consuming? If I didn't remember wrong, Kafka allows consumers to read `__consumer_offsets` and some admin tools are implemented by this way.

OK, the consume may expose to users.

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/InnerTopicProtectionTest.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+            producer.send(new ProducerRecord<>(topic, value), ((metadata, exception) -> {
+                if (assertException) {
+                    Assert.assertNotNull(exception);
+                } else {
+                    Assert.assertNull(exception);
+                }
+            })).get();
+        } catch (Exception e) {
+
+        }
```
Comment (BewareMyPower) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/InnerTopicProtectionTest.java:174]:
```suggestion
        try {
            producer.send(new ProducerRecord<>(topic, value)).get();
        } catch (Exception e) {
            if (assertException) {
                Assert.assertEquals(e.getCause().getMessage(),
                        "The request attempted to perform an operation on an invalid topic.");
            } else {
                Assert.fail();
            }
        }
```

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/InnerTopicProtectionTest.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    protected void resetConfig() {
+        offsetsTopicNumPartitions = 16;
+        conf = resetConfig(
+            brokerPort,
+            brokerWebservicePort,
+            kafkaBrokerPort);
+
+        log.info("Ports --  broker: {}, brokerWeb:{}, kafka: {}",
+            brokerPort, brokerWebservicePort, kafkaBrokerPort);
+    }
```
Comment (BewareMyPower) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/InnerTopicProtectionTest.java:101]: Could you explain why overwrite `resetConfig` here? I tried to remove the `resetConfig` code, and the tests still passed.

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/PendingProduce.java [hunk 3]
```diff
@@ -126,6 +126,12 @@ public void publishMessages() {
             throw new RuntimeException(e);
         }
 
+        if (persistentTopic.isSystemTopic()) {
+            log.error("Not support produce message to system topic: {}", persistentTopic);
+            responseFuture.complete(new PartitionResponse(Errors.INVALID_TOPIC_EXCEPTION));
+            return;
+        }
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/PendingProduce.java:133]: I noticed this case was not protected by test case. If there's something wrong with system topics, you can open a new PR to handle it.

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/InnerTopicProtectionTest.java (hunk 2 above)
Comment (hangc0276) [reply tests/src/test/java/io/streamnative/pulsar/handlers/kop/InnerTopicProtectionTest.java:101]: The test can cover turn on system topic and topic policy with resetConfig. However, KOP can't work when turn on topic policy. In the end, it turn off the topic policy in resetConfig. We can turn on when it fixed.

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/PendingProduce.java (hunk 3 above)
Comment (hangc0276) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/PendingProduce.java:133]: When turn on system topic, the KOP doesn't work. I turn off it in the test.

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/InnerTopicProtectionTest.java (hunk 2 above)
Comment (BewareMyPower) [reply tests/src/test/java/io/streamnative/pulsar/handlers/kop/InnerTopicProtectionTest.java:101]: So I think these two `resetConfig` could be removed as well.

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/InnerTopicProtectionTest.java (hunk 2 above)
Comment (BewareMyPower) [reply tests/src/test/java/io/streamnative/pulsar/handlers/kop/InnerTopicProtectionTest.java:101]: OK

**Comment by BewareMyPower** [comment]
I see it works well with system topic enabled. Is https://github.com/apache/pulsar/pull/10341 necessary?

**Comment by BewareMyPower** [comment]
Sorry I just misunderstood.

**Review by BewareMyPower — APPROVED** [review]

**Pull request merged**